File: src/main/java/com/example/demo/controller/UserSettingsController.java
package com.example.demo.controller;

import com.example.demo.dto.SaveUserSettingsRequest;
import com.example.demo.model.UserSettings;
import com.example.demo.service.UserSettingsService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.time.LocalTime;

@RestController
@RequestMapping("/api/user-settings")
@Validated
@RequiredArgsConstructor
public class UserSettingsController {

    private final UserSettingsService userSettingsService;


    @GetMapping
    public UserSettings getSettings(Authentication auth) {
        return userSettingsService.getSettings(auth.getName());
    }

    @PostMapping
    public UserSettings saveSettings(Authentication auth, @Valid @RequestBody SaveUserSettingsRequest body) {
        UserSettings settings = new UserSettings();
        settings.setUserName(auth.getName());
        String companyAffiliation = body.companyAffiliation();
        if (companyAffiliation != null && !companyAffiliation.isBlank())
            settings.setCompanyAffiliation(companyAffiliation);
        if (body.section() != null) {
            settings.setSection(body.section());
        }
        String branchOffice = body.branchOffice();
        if (branchOffice != null && !branchOffice.isBlank()) settings.setBranchOffice(branchOffice);
        if (body.workGroup() != null) {
            settings.setWorkGroup(body.workGroup());
        }
        String employeeNumber = body.employeeNumber();
        if (employeeNumber != null && !employeeNumber.isBlank()) settings.setEmployeeNumber(employeeNumber);
        LocalTime siteRegularHours = body.siteRegularHours();
        if (siteRegularHours != null) {
            settings.setSiteRegularHours(siteRegularHours);
        }
        String displayName = body.displayName();
        settings.setDisplayName(displayName != null && !displayName.isBlank() ? displayName : auth.getName());
        return userSettingsService.saveOrUpdate(settings);
    }

}

File: src/main/kotlin/com/example/demo/controller/UserSettingsController.kt
package com.example.demo.controller

import com.example.demo.model.UserSettings
import com.example.demo.service.UserSettingsService
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.*
import java.time.LocalTime

@RestController
@RequestMapping("/api/user-settings")
class UserSettingsController(private val userSettingsService: UserSettingsService) {

    /**
     * ユーザー設定を取得
     */
    @GetMapping
    fun getSettings(auth: Authentication): UserSettings? {
        return userSettingsService.getSettings(auth.name)
    }

    /**
     * ユーザー設定を保存または更新
     */
    @PostMapping
    fun saveSettings(auth: Authentication, @RequestBody body: Map<String, String?>): UserSettings {
        val settings = UserSettings(
            userName = auth.name,
            companyAffiliation = body["companyAffiliation"]?.takeIf { it.isNotBlank() },
            section = body["section"]?.takeIf { it.isNotBlank() }?.toIntOrNull(),
            branchOffice = body["branchOffice"]?.takeIf { it.isNotBlank() },
            workGroup = body["workGroup"]?.takeIf { it.isNotBlank() }?.toIntOrNull(),
            employeeNumber = body["employeeNumber"]?.takeIf { it.isNotBlank() },
            siteRegularHours = body["siteRegularHours"]?.takeIf { it.isNotBlank() }?.let { LocalTime.parse(it) },
            displayName = body["displayName"]?.takeIf { it.isNotBlank() } ?: auth.name
        )

        return userSettingsService.saveOrUpdate(settings)
    }
}
